bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java

```diff
@@ -943,9 +943,6 @@ public Iterable<Long> getActiveLedgersInRange(long firstLedgerId, long lastLedge
 
     @Override
     public void updateEntriesLocations(Iterable<EntryLocation> locations) throws IOException {
-        // Trigger a flush to have all the entries being compacted in the db storage
-        flush();
-
         entryLocationIndex.updateLocations(locations);
```

Comment:
How this change affects data consistency in case of e.g. node crash after updateLocations()?
As I understand, bookie can end up with index persisted and pointing to the data that hasn't been persisted.

Reply:
Looking back at the whole transaction steps,  Flush the current ledger storage write cache may want to ensure the remaining data has been flushed into the entry log file before updating the entres' index into RocksDB. But both Transaction Compactor and EntryLog Compactor are triggered flush operations to ensure those remaining data are flushed into the entry log file. So I think the Flush of the current ledger storage step is unnecessary and it will bring more throughput impact.

Reply:
For the standard compactor we are calling `EntryLogger.flush() `here before calling `updateEntriesLocations`
https://github.com/apache/bookkeeper/blob/405e72acf42bb1104296447ea8840d805094c787/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogCompactor.java#L120

My understanding is that with this change we are not flushing the index on RocksDB before actually updating the locations and we will wait for the next flush to happen.

In this [place](https://github.com/apache/bookkeeper/blob/ceba60565cf7cb438e9be4ab7416a2808b9168a1/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndex.java#L183) we flush the batch of writes related to the new entry locations passed to `updateEntriesLocations`

I wonder if we could risk to write these updates and then at the next flush overwrite the locations for the same entries with old data accumulated in `writeCacheBeingFlushed` https://github.com/apache/bookkeeper/blob/f5455f01584b1b0a592f020eed49d3cb774da0a9/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java#L817

Reply:
In fact, the data accumulated in `writeCacheBeingFlushed` is new data, because we are compacting the old data.

Refer to:
> #### Behavior Change
> After we removed the `flush the current ledger storage write cache` step, it brings one behavior change. https://github.com/apache/bookkeeper/blob/c765aea600fde1a8ac7cb8a53a1157a372026894/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/SingleDirectoryDbLedgerStorage.java#L947
> Let me take one case for example.
> Timeline: [t1, t2, t3]
t1: One entry (LedgerId = 1, EntryId = 1) with value "test-v1" written into the bookie from one bookie client
t2: The entry (LedgerId = 1, EntryId = 1) with new value "test-new-v1" written into the bookie from one bookie client
t3: Bookie triggers compaction to compact the entry written in t1, the entry (LedgerId = 1, EntryId = 1) with value "test-v1" will be written into the current entry log file, and update the entry's lookup index.
> ##### Before this change
> Due to the compactor flushing the current ledger storage write cache before updating the entry's lookup index, the updated value of the entry "test-new-v1" in `t2` will be flushed into storage and removed from the write cache. If we get the entry (LedgerId = 1, EntryId = 1) after `t3`, we will get the old value `test-v1`
> ##### After this change
> Due to this PR removed the compactor flushing the current ledger storage write cache, it has two cases:
> - If `t3` happens when `t2`'s updated data is still located in the ledger storage's write cache, the new data updated in `t2` will override the old data written in t3. When we get the entry (LedgerId = 1, EntryId = 1), we will get the new value `test-new-v1`
> - If `t3` happens when `t2`'s updated data has been flushed into the entry log file and removed from the write cache, the old entry written in `t3` will override the new entry written in `t2`. When we get the entry (LedgerId = 1, EntryId=1), we will get the old value `test-v1`.
> IMO, we should always return the new value `test-new-v1`, not the old value `test-v1`. If we need to make sure getting the entry always returns the new value, we need more checks in writing the old value in the compaction stage.
> In Pulsar's general case, updating the entry's value won't happen.

Reply:
Agree with Hang.

> If t3 happens when t2's updated data has been flushed into the entry log file and removed from the write cache, the old entry written in t3 will override the new entry written in t2. When we get the entry (LedgerId = 1, EntryId=1), we will get the old value test-v1.

In this case. We should check if the remaining data's new location is overridden by the client incoming entries.

Reply:
> Agree with Hang.

+1 

> > If t3 happens when t2's updated data has been flushed into the entry log file and removed from the write cache, the old entry written in t3 will override the new entry written in t2. When we get the entry (LedgerId = 1, EntryId=1), we will get the old value test-v1.
> 
> In this case. We should check if the remaining data's new location is overridden by the client incoming entries.

In this case, we should compare whether the location of the entry in the compression entrylog is consistent with the location in the current EntryLocationIndex?